py3status/modules/updates.py

```diff
@@ -0,0 +1,199 @@
+# -*- coding: utf-8 -*-
+"""
+Display numbers of updates for various linux distributions.
+
+Configuration parameters:
+    cache_timeout: refresh interval for this module (default 600)
+    format: display format for this module, otherwise auto (default None)
+    thresholds: specify color thresholds to use *(default [(0, 'darkgray'),
+        (10, 'degraded'), (20, 'orange'), (30, 'bad')])*
+
+Format placeholders:
+    {update}  number of updates, eg 0
+    {apk}     number of updates, eg 0 .. Alpine Linux     .. NOT TESTED
+    {apt}     number of updates, eg 0 .. Debian, Ubuntu
+    {auracle} number of updates, eg 0 .. Arch Linux (AUR)
+    {eopkg}   number of updates, eg 0 .. Solus
+    {pacman}  number of updates, eg 0 .. Arch Linux
+    {xbps}    number of updates, eg 0 .. Void Linux       .. NOT TESTED
+    {yay}     number of updates, eg 0 .. Arch Linux (AUR)
+    {zypper}  number of updates, eg 0 .. openSUSE         .. NOT TESTED
+
+Color thresholds:
+    xxx: print a color based on the value of `xxx` placeholder
+
+@author lasers
+
+Examples:
+```
+# multiple distributions, same format
+updates {
+    format = '[\?not_zero Updates [\?color=update {update}]]'
+}
+
+# Arch Linux
+updates {
+    format = 'Updates [\?color=pacman {pacman}]/[\?color=auracle {auracle}]'
+    # format = 'Updates [\?color=pacman {pacman}]/[\?color=yay {yay}]'
+}
+```
+
+SAMPLE OUTPUT
+[{'full_text': 'Apk '}, {'full_text': '3', 'color': '#a9a9a9'}]
+
+14pkgs
+[{'full_text': 'Apt '}, {'full_text': '14', 'color': '#00FF00'}]
+
+29and5pkgs
+[{'full_text': 'Pacman '}, {'full_text': '29 ', 'color': '#FFFF00'},
+{'full_text': 'Auracle '}, {'full_text': '5', 'color': '#a9a9a9'}]
+
+35pkgs
+[{'full_text': 'Zypper '}, {'full_text': '34', 'color': '#ffa500'}]
+
+45pkgs
+[{'full_text': 'Xbps '}, {'full_text': '45', 'color': '#FF0000'}]
+
+no_updates
+{'full_text': 'No Updates'}
+"""
+
+
+class Update:
+    def __init__(self, parent, name, command):
+        self.parent = parent
+        self.name = name
+        self.command = command
+
+    def get_output(self):
+        try:
+            return self.parent.py3.command_output(self.command)
+        except self.parent.py3.CommandError:
+            return None
+
+    def count_updates(self, output):
+        return len(output.splitlines())
+
+    def get_updates(self):
+        output = self.get_output()
+        if output is None:
+            return {self.name: output}
+        return {self.name: self.count_updates(output)}
+
+
+class Apt(Update):
+    def count_updates(self, output):
+        return len(output.splitlines()[1:])
```

Comment:
That can work. What's the replacement code for `Eopkg` and `Zypper`? @tobes Thoughts on removing or keeping classes?

I'm icky on adding `notifications` support for `updates`. I think nobody but you would use this.

Reply:
I'm eyeballing my return code. No actual output here.

Eopkg prints `No packages to upgrade.`. Detect that and print `0`... Otherwise count lines.

It looks like Zypper prints 4 lines of headers, then an empty line, then comes a list of updates. Count them. Maybe it's just `[5:]` instead of 4 headers and an empty line, but I don't have an output here.... I can't say anything solid about Zypper.

I'd suggest you and @cyrinux to stick with `external_script` for notification support. I'm not sure if we want `updates` to accept custom commands. Also, `managers` config is supposed to be hidden. Users are expected to use `{pacman}` to get `{pacman}` alone (even with yay installed).

Reply:
Why not allow accepting custom commands? You see how long it takes to merge PRs, what if tomorrow a new package manager gets popular, but we will not allow people to use them, and will tell them to switch to `external_script`? If that's the case, then let's tell everyone to use external_script.

> Eopkg prints No packages to upgrade

`| grep -v 'No packages to upgrade'`

Reply:
>  @tobes Thoughts on removing or keeping classes?

Classes are great, OO code can be really powerful.  However sometimes people get carried away by the idea, and this can make code overly complex.  For me I have a few questions I always ask.  Here's one.

Do we have instance data that we need to keep?

Yes, then a class may be good.
No, hmm do we need a class? could we just have a function?

So here it feels maybe we don't need the class.

For me what is interesting is if we can just have simple commands as suggested by @maximbaz.  If the answer is yes then a dict is all we need.  I still have a disagreement with how this module currently is because it is creating work for users.

Reply:
I'll try my best to answer them. :-)

>Do we have instance data that we need to keep?

We're keeping `name`, `command`. Some update managers does not need modifications and will print out perfect output. I have a `update-manager` list outside to check everything first. The `name` is used to return `{self.name: count}`.

There are 3 functions. One to fix the output. One to offset or manipulate lines. One deals with putting things together (as well as deal with offline `None` if needed for some users). Cower errors on success (cower is apparently obsolete since Feb). It's all small tweaks.

>I still have a disagreement with how this module currently is because it is creating work for users.

Can you expand more on that? All users should do is to add `order += "updates"` then it'll figure out things automatically. If users want to customize, then they can make their own `format` with either `{update}` or explicit placeholders. @maximbaz added three different commands to (hidden) `managers and it prints out okay.

Reply:
Undocumented features are bad, don't hide configuration via `managers` pls

Reply:
For cower you can use `|| true`

Reply:
>Undocumented features are bad

Debug options. Also, #1519

Reply:
>  >   Do we have instance data that we need to keep?
>
> We're keeping name, command

So we do not have state - So I think a class is the wrong approach here.